### ISSUE TITLE
Custom Request-Id for client side requests

### ISSUE THREAD

**Issue by emanuel-v-r** [dotnet/aspnetcore#38850]
On every new request, aspnetcore creates a new Activity and sets a RequestId, TraceId, and SpanId, and spans it through the logs.
Is it possible to somehow set the ParentId to a UUID coming from the client-side (javascript) via a custom HTTP header?
I understand that if this UUID comes in the Request-Id HTTP header it automatically uses it, but what if it comes in a different custom header?
Thank you.

**Comment by davidfowl** [comment]
Client as in JavaScript or .NET client?

**Comment by emanuel-v-r** [comment]
> Client as in JavaScript or .NET client?

Javascript client, from the UI. This way we would be able to track down the whole path of a request coming from the browser.
As I understand Request-Id header is already automatically added to requests on the .NET side (at least when using HttpClientFactory).

**Comment by adityamandaleeka** [comment]
@Tratcher @noahfalk This should be possible now right?

**Comment by Tratcher** [comment]
It can be customize, I just need to find the example.

Here's the code in hosting, it uses a DistributedContextPropagator.
https://github.com/dotnet/aspnetcore/blob/73db5e40bbdeb4f04e5e05b9090f253760d1e1ee/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L286-L292

The only reference I can find is https://devblogs.microsoft.com/dotnet/announcing-net-6-preview-7/#libraries-system-diagnostics-propagators, but @noahfalk would know better.

**Comment by cijothomas** [comment]
> On every new request, aspnetcore creates a new Activity and sets a RequestId, TraceId, and SpanId, and spans it through the logs. Is it possible to somehow set the ParentId to a UUID coming from the client-side (javascript) via a custom HTTP header? I understand that if this UUID comes in the Request-Id HTTP header it automatically uses it, but what if it comes in a different custom header? Thank you.

This should be possible custom "propagators" (https://github.com/dotnet/runtime/issues/50658). The default propagator respects the W3CTraceContext spec, i.e it looks for "traceparent" header (with a fallback to "request-id" to maintain back compact with legacy header).

**Comment by sebastienros** [comment]
> This should be possible custom "propagators"

@cijothomas do you have any pointers to some documentation?

**Comment by sebastienros** [comment]
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md

**Comment by cijothomas** [comment]
> > This should be possible custom "propagators"
> 
> @cijothomas do you have any pointers to some documentation?

Nothing other than the design proposal : https://github.com/dotnet/runtime/issues/50658#issuecomment-879212250

**Comment by emanuel-v-r** [comment]
Thank you all for the help here, I will analyze the source code in order to understand how to provide a custom propagator. Some documentation would be nice to have though.

**Comment by noahfalk** [comment]
@tarekgh @MihaZupan - did any documentation ever get created for the propagators feature? For example it would be ideal to have it documented together with our existing [distributed tracing](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing) docs.

In the meantime if there are no docs I think the short story is roughly:
1. Implement a custom version of DistributedContextPropagator. You can use an [existing implementation](https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/LegacyPropagator.cs,10) as a guide. If you want to change how new requests have their id populated then you want to change the implementation of [ExtractTraceIdAndState](https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/LegacyPropagator.cs,45). For example if you want the id to be read from a header called "MySpecialIdHeader" and there is no encoding for tracestate you could write:

```C#
public override void ExtractTraceIdAndState(object? carrier, PropagatorGetterCallback? getter, out string? traceId, out string? traceState)
        {
            if (getter is null)
            {
                traceId = null;
                traceState = null;
                return;
            }
            getter(carrier, "MySpecialIdHeader", out traceId, out _);
            traceState = null;
        }
```

If you need to handle the new header in addition to messages that use other more typical headers you can also consider falling back to check other headers when this one isn't present.

Also update the Fields property to enumerate any header names you might want to look up per-request.

2. Assign a new instance of your propagator to the static property DistributedContextPropagator.Current when your app starts up. This needs to happen before it is captured by the WebHostBuilder [here](https://source.dot.net/#Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostBuilder.cs,91).

**Comment by tarekgh** [comment]
> did any documentation ever get created for the propagators feature? For example it would be ideal to have it documented together with our existing distributed tracing docs.

We have documented the APIs [here](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.distributedcontextpropagator?view=net-6.0). Having a conceptual doc for propagators is a good idea too.

**Comment by adityamandaleeka** [comment]
Reading through this now, I don't think there's any work in ASP.NET Core for this so I'll close the issue.

@tarekgh You may want to file a tracking item in runtime's docs for the conceptual doc on propagators.
